packages/protocol/contracts/signal/SignalService.sol

```diff
@@ -140,13 +150,25 @@ contract SignalService is AuthorizableContract, ISignalService {
 
             uint64 hopChainId = uint256(label).toUint64();
 
-            verifyMerkleProof(
-                stateRoot, hopChainId, hop.relay, hop.stateRoot, hop.merkleProof
+            verifiedStorageRoots[hopChainId][hop.relay][stateRoot] = verifyMerkleProof(
+                stateRoot,
+                hopChainId,
+                hop.relay,
+                hop.stateRoot,
+                hop.merkleProof,
+                verifiedStorageRoots[hopChainId][hop.relay][stateRoot]
             );
             stateRoot = hop.stateRoot;
         }
 
-        verifyMerkleProof(stateRoot, srcChainId, app, signal, p.merkleProof);
+        verifiedStorageRoots[srcChainId][app][stateRoot] = verifyMerkleProof(
```

Comment:
If we cache all stateRoot along the multi-hop path, it may be quite expensive. I'm not sure how often a single state root will be reused by multiple signals, we may end up with very expensive bridging experience.

Reply:
@adaki2004 may I advise you to revert to the previous commit and work on caching with a separate PR?

Reply:
> If we cache all stateRoot along the multi-hop path, it may be quite expensive. I'm not sure how often a single state root will be reused by multiple signals, we may end up with very expensive bridging experience.

I do think the other way around.
Currently bridging costs 410-440K gas for ether and 520K for token bridging. This is an extra 20K (5%) per only one time per state root, and a relayer (if implemented it that way, which can be done easily) could make an aggregation off-chain by the following:

Bob wants to bridge at height 99.
Cecile wants to bridge at height 100.
Dave wants to birdge at height 101.

So the relayer relays over 1 message at WorldStateRoot 102 first e.g.: Bob's, and then for Cecile and Dave this is cheap because of caching. The whole point of caching is optimization and making things more effective.

I'd wait for @Brechtpd take before doing any reverts in the commit.

Reply:
As discussed offline i reverted the commit and opened up a seperate PR just about caching, [here](https://github.com/taikoxyz/taiko-mono/pull/15712/).

Reply:
I think we should also cache all the hops. Let's take for example L1 -> L2 and L2 -> L2. In both cases, these need to get the signal root from the L1 signal contract. So in the L1 -> L2 we would now cache and reuse it as much as possible. However, we need the exact same data for the L2 -> L2 case, so it would be inefficient to not at least check if we have the L1 signal root cached for the L1 signal contract. But then, why not simply also cache it in it the L2 -> L2? Then it will also became usable for the L1 -> L2 signals.

So caching the different hops seems much more efficient in the L2 -> L2 case, and it should also make the L1 -> L2 bridging cheaper because now cached values from the L2 -> L2 can now also be used.

Worst case scenario for somebody doing L2 -> L2 bridging would indeed be an extra 40k gas, but because it's likely that some L1 signal root can be reused already for the first hop, it's likely to already be much cheaper in most practical use cases.

Users/smart relayers that optimize for bridging cost can just wait a bit until somebody else was willing to pay the initial account proof cost, so really that should be very infrequent. Bridging already takes at least 24 hours anyways, waiting a bit and being able to do all bridging transactions each hour or something with only the cost of a single account proof seems very reasonable.

Reply:
@dantaik any objections ?
If not i'd put it back on the other PR, here: https://github.com/taikoxyz/taiko-mono/pull/15712

Reply:
Brecht, the concern I have is not uncommon: caching is not cost free and we shall cache things if they will be used such that the overcall gas savings justify the cost.

Reply:
For sure! But caching should be pretty much always be more efficient for users. The only case where I think this may be more expensive if a user does the bridging on its own and wants to do it as fast as possible so no cached root can be used. This seems like a niche case to me, and the extra cost in this case is even small compared to the full cost (10% per hop in the worst case where no cached state can be used?).

I assume that the normal use case would be that some relayer handles most bridge transactions. This relayer has full control over which storage root to use to do the Merkle proof. So that means that for a relayer, things get already much more efficient as long as he needs to do only 2 bridge transactions! He just needs to make sure he submits all bridge transactions against the same state root, and then he is sure he only needs to pay for a single account Merkle proof and a single cache!

I believe this is even more efficient than the previous method of always caching the signal root/state root in the protocol (https://github.com/taikoxyz/taiko-mono/pull/13214#issuecomment-1445381454).